text/0000-default-fields.md

```diff
@@ -0,0 +1,305 @@
+
+The value of a field default must be a compile-time expression. So any expression that's valid as a `const` can be used as a field default. This ensures values that aren't specified by the caller are deterministic and cheap to produce. A type doesn't need to derive `Default` to be valid as a field default.
+
+Valid:
+
+```rust
+struct Foo {
+    a: &'static str,
+    b: bool = true,
+    c: i32,
+}
+```
+
+Invalid:
+
+```rust
+struct Foo {
+    a: &'static str,
+    b: Vec<bool> = Vec::new(),
+                   ^^^^^^^^^^
+                   // error: calls in field defaults are limited to struct and enum constructors
+    c: i32,
+}
+```
+
+The above error is based on `E0015` for trying to initialise a constant with a non-constant expression. As the scope of constant expressions changes this message will change too.
+
+Field defaults are like a shorthand for the 'real' initialiser, where values for missing fields are added with the supplied default expression:
+
+```rust
+let foo = Foo {
+    a: "Hello",
+    c: 42,
+};
+```
+
+is equivalent to:
+
+```rust
+let foo = Foo {
+    a: "Hello",
+    b: true,
+    c: 42,
+};
+```
+
+The mechanism isn't exactly a shorthand because the structure can still be initialised using field defaults even if `b` is private. The caller still can't interact with private fields directly so privacy isn't violated.
+
+When a caller doesn't supply a field value during initialisation and there is no default available then the `E0063` missing field error applies.
+
+## Order of precedence
+
+Supplied field values take precedence over field defaults:
+
+```rust
+// `b` is `false`, even though the field default is `true`
+let foo = Foo {
+    a: "Hello",
+    b: false,
+    c: 42,
+};
+```
+
+Supplied field values in functional updates take precedence over field defaults:
+
+```rust
+// `b` is `false`, even though the field default is `true`
+let foo = Foo {
+    a: "Hello",
+    c: 0,
+    ..Foo {
+        a: "Hello",
+        b: false,
+        c: 0
+    }
+};
+```
+
+## Deriving `Default`
+
+When deriving `Default`, supplied field defaults are used instead of the type default. This is a feature of `#[derive(Default)]`.
+
+```rust
+#[derive(Default)]
+struct Foo {
+    a: &'static str,
+    b: bool = true,
+    c: i32,
+}
+
+// `b` is `true`, even though `bool::default()` is `false`
+let foo = Foo::default();
+```
+
+Field defaults allow `#[derive(Default)]` to be used more widely because the types of fields with default values don't need to implement `Default`.
+
+## Enabling backwards compatibility
+
+With no special syntax, additional fields can be added to a struct in a non-breaking fashion. Say we have the following API and consumer:
+
+```rust
+mod data {
+    pub struct Foo {
+        pub a: &'static str,
+        pub c: i32,
+    }
+}
+
+let foo = data::Foo {
+    a: "Hello",
+    c: 42,
+}
+```
+
+We can add a new field `b` to this `struct` with a default value, and the calling code doesn't change:
```

Comment:
My immediate thought is that `..` in patterns would also have to be optional then for consistency. I'm cautious about that though because it might increase the  scope of this thing too much.

The other option is make `..` required for structs with defaults, or stop pretending it's always backwards compatible.